### ISSUE TITLE
SingularField rule incorrectly reported violated with switch in method that both assigns and reads field

### ISSUE THREAD

**Issue by theneva** [pmd/pmd#174]
I have the following JAX-RS event listener that uses a Prometheus Summary to measure the time of all non-404 responses. It violates the [java/design/SingularField rule](https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/design.html#SingularField).

```java
// package, imports

@Provider
public class LatencyRequestEventListener implements RequestEventListener {
    private static final Summary SUMMARY_LATENCY_SECONDS = …;

    private final ResourceInfo resourceInfo;

    private Summary.Timer timer; // BREAKS HERE

    LatencyRequestEventListener(final ResourceInfo resourceInfo) {
        this.resourceInfo = resourceInfo;
    }

    @Override
    public void onEvent(final RequestEvent event) {
        switch (event.getType()) {
            case RESOURCE_METHOD_START:
                final ResourceMethod matchedResourceMethod = event.getUriInfo().getMatchedResourceMethod();

                if (matchedResourceMethod == null) {
                    return;
                }

                timer = SUMMARY_LATENCY_SECONDS // ASSIGN FIELD
                    .labels(
                        matchedResourceMethod.getHttpMethod()
                        , resourceInfo.getResourceClass().getSimpleName() + "." + resourceInfo.getResourceMethod().getName()
                    )
                    .startTimer();
                break; // Don't keep going (both cases cannot be reached in the same call)
            case FINISHED:
                if (timer != null) {
                    timer.observeDuration(); // READ FIELD
                }
        }
    }
}
```

This results in the following error:

> /path/to/project/…/listeners/LatencyRequestEventListener.java:25: Perhaps 'timer' could be replaced by a local variable.

`timer` can't be a local variable, so I don't think it's correct to report this rule violated.

I understand that the rule is violated if a field is only accessed by a single method. I also acknowledge that having the same method be called twice to first assign a field (the `RESOURCE_METHOD_START` case) and then read the field (`FINISHED`) may not be a good design. However, that's how the JAX-RS people designed the `RequestEventListener` class.

If I pull the code for the `RESOURCE_METHOD_START` case into a separate method, it passes with no problem since multiple methods access the field.

Is this a bug, or am I doing something wrong?

**Comment by jsotuyod** [comment]
Hi @theneva and thanks for your report!

You are absolutely right, this is a false positive, mostly down to the fact PMD can't tell how calls will be performed in time.

However, I believe we could improve this scenario by checking, not only that the field is only being written / read on a single method, but is actually being written in all branches before being read (and therefore not relying on previously set values from other calls). This however, would probably need to rely on our Data Flow Analysis framework, [which is in serious need of a revamp](https://github.com/pmd/pmd/wiki/Project-Ideas#data-flow-analysis).

This is definitely in our backlog, but will take some time to get fixed. In the meantime you can either [suppress the warning locally](https://pmd.github.io/pmd-5.5.2/usage/suppressing.html) and document it as a false positive with a comment even linking back to this issue if you would like so.

**Comment by theneva** [comment]
Hey @jsotuyod, thanks for the reply! I just realised I was supposed to file the bug report on SourceForge, so I did that: https://sourceforge.net/p/pmd/bugs/1558/

I'll close this issue here. Thanks again :)

**Comment by jsotuyod** [comment]
@theneva don't worry, we have decided to move away from SourceForge issues and into GitHub. We will eventually move all issues here anyway.

**Comment by theneva** [comment]
Oh, cool! I'll leave it open in both places then.

**Comment by adangel** [comment]
This has been fixed in PMD 7.
